File: frontend/src/services/prompt.ts
import {
  CreatePrompt,
  DeletePrompt,
  GetPrompt,
  ListPrompts,
  RenderPrompt,
  UpdatePrompt,
} from "@/wailsjs/wailsjs/go/app/App";
import { dto } from "@/wailsjs/wailsjs/go/models";
import { unwrapMany, unwrapOne, unwrapString } from "./utils";

export interface Prompt {
  id: number;
  name: string;
  systemPrompt: string;
  userPrompt: string;
  placeholders: string[];
  createdAt: string;
  updatedAt: string;
}

export interface PromptRenderResult {
  system: string;
  user: string;
}

export function mapPrompt(x: dto.Prompt): Prompt {
  return {
    id: x.id,
    name: x.name,
    systemPrompt: x.systemPrompt,
    userPrompt: x.userPrompt,
    placeholders: x.placeholders,
    createdAt: x.createdAt,
    updatedAt: x.updatedAt,
  };
}

export function mapPromptRenderResult(x: dto.PromptRenderResult): PromptRenderResult {
  return { system: x.system, user: x.user };
}

export async function listPrompts(): Promise<Prompt[]> {
  const res = await ListPrompts();
  return unwrapMany<dto.Prompt>(res).map(mapPrompt);
}

export async function getPrompt(id: number): Promise<Prompt> {
  const res = await GetPrompt(id);
  return mapPrompt(unwrapOne<dto.Prompt>(res));
}

export async function createPrompt(input: Omit<Prompt, "id" | "createdAt" | "updatedAt">): Promise<string> {
  const payload = new dto.Prompt({
    name: input.name,
    systemPrompt: input.systemPrompt,
    userPrompt: input.userPrompt,
    placeholders: input.placeholders,
  });
  const res = await CreatePrompt(payload);
  return unwrapString(res);
}

export async function updatePrompt(input: Omit<Prompt, "createdAt" | "updatedAt">): Promise<string> {
  const payload = new dto.Prompt({
    id: input.id,
    name: input.name,
    systemPrompt: input.systemPrompt,
    userPrompt: input.userPrompt,
    placeholders: input.placeholders,
  });
  const res = await UpdatePrompt(payload);
  return unwrapString(res);
}

export async function deletePrompt(id: number): Promise<string> {
  const res = await DeletePrompt(id);
  return unwrapString(res);
}

export async function renderPrompt(promptId: number, variables: Record<string, string>): Promise<PromptRenderResult> {
  const res = await RenderPrompt(promptId, variables);
  return mapPromptRenderResult(unwrapOne<dto.PromptRenderResult>(res as any));
}

File: internal/app/prompt.go
package app

import (
	"Postulator/internal/domain/entities"
	"Postulator/internal/dto"
	"Postulator/pkg/errors"
	"context"
)

// Prompt bindings

func (a *App) CreatePrompt(p *dto.Prompt) *dto.Response[string] {
	if p == nil {
		return dtoErr[string](errors.Validation("prompt payload is required"))
	}
	entity := &entities.Prompt{ID: p.ID, Name: p.Name, SystemPrompt: p.SystemPrompt, UserPrompt: p.UserPrompt}
	if err := a.promptSvc.CreatePrompt(context.Background(), entity); err != nil {
		return dtoErr[string](asAppErr(err))
	}
	return &dto.Response[string]{Success: true, Data: "created"}
}

func (a *App) GetPrompt(id int64) *dto.Response[*dto.Prompt] {
	pr, err := a.promptSvc.GetPrompt(context.Background(), id)
	if err != nil {
		return dtoErr[*dto.Prompt](asAppErr(err))
	}
	return &dto.Response[*dto.Prompt]{Success: true, Data: dto.FromPrompt(pr)}
}

func (a *App) ListPrompts() *dto.Response[[]*dto.Prompt] {
	items, err := a.promptSvc.ListPrompts(context.Background())
	if err != nil {
		return dtoErr[[]*dto.Prompt](asAppErr(err))
	}
	return &dto.Response[[]*dto.Prompt]{Success: true, Data: dto.FromPrompts(items)}
}

func (a *App) UpdatePrompt(p *dto.Prompt) *dto.Response[string] {
	if p == nil {
		return dtoErr[string](errors.Validation("prompt payload is required"))
	}
	entity := &entities.Prompt{ID: p.ID, Name: p.Name, SystemPrompt: p.SystemPrompt, UserPrompt: p.UserPrompt}
	if err := a.promptSvc.UpdatePrompt(context.Background(), entity); err != nil {
		return dtoErr[string](asAppErr(err))
	}
	return &dto.Response[string]{Success: true, Data: "updated"}
}

func (a *App) DeletePrompt(id int64) *dto.Response[string] {
	if err := a.promptSvc.DeletePrompt(context.Background(), id); err != nil {
		return dtoErr[string](asAppErr(err))
	}
	return &dto.Response[string]{Success: true, Data: "deleted"}
}

func (a *App) RenderPrompt(promptID int64, placeholders map[string]string) *dto.Response[*dto.PromptRenderResult] {
	sys, user, err := a.promptSvc.RenderPrompt(context.Background(), promptID, placeholders)
	if err != nil {
		return dtoErr[*dto.PromptRenderResult](asAppErr(err))
	}
	return &dto.Response[*dto.PromptRenderResult]{Success: true, Data: &dto.PromptRenderResult{System: sys, User: user}}
}
